### PR TITLE
feat: wf pod restart

### LINKED ISSUE #4061: Feature: auto retrigger ci/pre-cd/post-cd pods if deleted during running state.
### 🔖 Feature description

This feature should re trigger ci/pre-cd/post-cd pods if they get deleted during there running state

### 🎤 Pitch

now ci/pre-cd/post-cd pods get deleted if node on which they scheduled goes away. so users have to manually retrigger respective workflows. it would be usefull if we can detect this deletion and retrigger them automatically.

### ✌️ Solution

re-trigger ci/pre-cd/post-cd pods if they get deleted during there running state

### 🔄️ Alternative

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#6122

### PR DESCRIPTION
**Pull request by gireesh-devtron**
# Description
This feature is to retrigger ci,pre-cd and post-cd workflows incase they get deleted before reaching terminal state.

Fixes #4061 
## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/pipeline/CdHandler.go [hunk 1]
```diff
@@ -170,6 +171,44 @@ const WorklowTypeDeploy = "DEPLOY"
...
+		}
+	}
+	retryCnt, err := impl.cdWorkflowRepository.FindRetriedWorkflowCountByReferenceId(runner.Id)
+	if err != nil {
+		impl.Logger.Errorw("error in FindRetriedWorkflowCountByReferenceId ", "err", err, "cdWorkflowRunnerId", runner.Id)
+		return
+	}
+
+	if retryCnt >= impl.cdConfig.MaxCdWorkflowRunnerRetries {
+		impl.Logger.Debugw("maximum retries for this workflow are exhausted, not re-triggering again", "retries", retryCnt, "wfrId", runner.Id)
```
Comment (kripanshdevtron) [inline pkg/pipeline/CdHandler.go:194]: convert this to Info level msg

File: pkg/pipeline/CiHandler.go [hunk 2]
```diff
@@ -1461,7 +1572,22 @@ func (impl *CiHandlerImpl) UpdateCiWorkflowStatusFailure(timeoutForFailureCiBuil
...
+						impl.Logger.Warnw("error in getting cluster Config", "err", err, "clusterId", env.Cluster.Id)
+						continue
+					}
+
+					client, err = impl.K8sUtil.GetCoreV1Client(clusterConfig)
+					if err != nil {
+						impl.Logger.Warnw("error in getting core v1 client using clusterConfig", "err", err, "clusterId", env.Cluster.Id)
+						continue
+					}
+					ns = env.Namespace
```
Comment (kripanshdevtron) [inline pkg/pipeline/CiHandler.go:1588]: create separate func for this

File: pkg/pipeline/CiHandler.go [hunk 3]
```diff
@@ -1473,12 +1599,30 @@ func (impl *CiHandlerImpl) UpdateCiWorkflowStatusFailure(timeoutForFailureCiBuil
...
-			ciWorkflow.Status = "Failed"
-			ciWorkflow.PodStatus = "Failed"
-			ciWorkflow.Message = "marked failed by job"
+			if isPodDeleted {
+				ciWorkflow.Status = "Failed"
+				ciWorkflow.Message = POD_DELETED_MESSAGE
+				if ciWorkflow.Status != WorkflowCancel {
+					retryCount, refCiWorkflow, err := impl.getRefWorkflowAndCiRetryCount(ciWorkflow)
+					if err != nil {
+						impl.Logger.Errorw("error in getRefWorkflowAndCiRetryCount", "ciWorkflowId", ciWorkflow.Id)
```
Comment (kripanshdevtron) [inline pkg/pipeline/CiHandler.go:1616]: error not logged here

Comment (gireesh-devtron) [reply pkg/pipeline/CiHandler.go:1616]: done

File: pkg/pipeline/CiHandler.go (hunk 2 above)
Comment (gireesh-devtron) [reply pkg/pipeline/CiHandler.go:1588]: done

File: pkg/pipeline/CdHandler.go (hunk 1 above)
Comment (gireesh-devtron) [reply pkg/pipeline/CdHandler.go:194]: done

File: internal/sql/repository/pipelineConfig/CdWorfkflowRepository.go [hunk 4]
```diff
@@ -486,6 +490,15 @@ func (impl *CdWorkflowRepositoryImpl) FindWorkflowRunnerById(wfrId int) (*CdWork
 	return wfr, err
 }
 
+func (impl *CdWorkflowRepositoryImpl) FindRetriedWorkflowCountByReferenceId(wfrId int) (int, error) {
+	retryCount := 0
+	query := fmt.Sprintf("select count(*) "+
```
Comment (kripanshdevtron) [inline internal/sql/repository/pipelineConfig/CdWorfkflowRepository.go:495]: use `id` instead of `*` here

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**